packages/contractkit/src/identity/metadata-manager.ts

```diff
@@ -0,0 +1,86 @@
+import { chain, Either, map } from 'fp-ts/lib/Either'
+import { pipe } from 'fp-ts/lib/pipeable'
+import { chain as andThen, fromEither, TaskEither } from 'fp-ts/lib/TaskEither'
+import {
+  asClassicPromise,
+  asThrowable,
+  fetchAsTaskEither,
+  NetworkError,
+  writeToFileSystem,
+} from '../utils/fp-ts'
+import {
+  AttestationServiceURLClaim,
+  Claim,
+  ClaimNotFoundError,
+  ClaimTypes,
+  findClaim,
+  IdentityMetadata,
+  NameClaim,
+  parseMetadata,
+  serializeMetadata,
+  SignedClaim,
+  validateMetadata,
+  ValidationError,
+} from './types'
+
+const now = () => Math.round(new Date().getTime() / 1000)
+
+export type MetadataError = NetworkError | ValidationError
+
+export class MetadataManager {
+  emptyIdentityMetadata: IdentityMetadata = {
+    claims: [],
+  }
+
+  fetchMetadataTE = (url: string): TaskEither<MetadataError, IdentityMetadata> =>
+    pipe(
+      fetchAsTaskEither(url),
+      andThen((body) =>
+        pipe(
+          body,
+          parseMetadata,
+          chain(validateMetadata),
+          fromEither
+        )
+      )
+    )
+
+  findAttestationServiceClaimE = (data: IdentityMetadata): Either<ClaimNotFoundError, string> => {
+    return pipe(
+      data,
+      findClaim(ClaimTypes.ATTESTATION_SERVICE_URL),
+      // @ts-ignore Typescript can't infer that the claim must be an Attestation service claim
```

Comment:
it can, the problem is in findClaim ;)

Reply:
I'm taking suggestions on how to improve `findClaim`. I tried, but couldn't figure out an implementation that correctly types the returned claim unfortunately. I maybe naively blame the compiler